samples/boards/sensortile_box_pro/ble-fw-upgrade/src/main.c

```diff
@@ -0,0 +1,378 @@
+/*
+ * Copyright (c) 2024 STMicroelectronics
+ *
+ * The BlueNRG-LP, BlueNRG-LPS UART bootloader protocol (AN5471)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <stdio.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/led.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ble_fw_upgrade, LOG_LEVEL_DBG);
+
+#define BLE_FLASH_START_ADDRESS 0x10040000
+
+/* Defined in ${ZEPHYR_HAL_ST_MODULE_DIR}/ble_firmware/ble_fw.c */
+extern const uint8_t ble_fw_img[];
+extern const int ble_fw_img_len;
+
+const struct gpio_dt_spec green_gpio = GPIO_DT_SPEC_GET(DT_ALIAS(led0), gpios);
+const struct gpio_dt_spec red_gpio = GPIO_DT_SPEC_GET(DT_ALIAS(led1), gpios);
+const struct gpio_dt_spec blue_gpio = GPIO_DT_SPEC_GET(DT_ALIAS(led3), gpios);
+
+typedef enum {
+	BLE_FW_UPG_START = 0x0,
+	BLE_FW_UPG_OK = 0x1,
+	BLE_FW_UPG_ERR = 0x2,
+} ble_fw_upg_status;
+
+static void led_pattern_out(ble_fw_upg_status status)
+{
+	int i;
+
+	gpio_pin_set_dt(&blue_gpio, 0);
+	gpio_pin_set_dt(&green_gpio, 0);
+	gpio_pin_set_dt(&red_gpio, 0);
+
+	switch (status) {
+	case BLE_FW_UPG_START:
+		for (i = 0; i < 7; i++) {
+			gpio_pin_toggle(blue_gpio.port, blue_gpio.pin);
+			k_sleep(K_MSEC(200));
+		}
+		break;
+
+	case BLE_FW_UPG_OK:
+		for (i = 0; i < 7; i++) {
+			gpio_pin_toggle(green_gpio.port, green_gpio.pin);
+			k_sleep(K_MSEC(200));
+		}
+		break;
+
+	case BLE_FW_UPG_ERR:
+		for (i = 0; i < 7; i++) {
+			gpio_pin_toggle(red_gpio.port, red_gpio.pin);
+			k_sleep(K_MSEC(200));
+		}
+		break;
+	}
+}
+
+static int led_init(void)
+{
+	/* led 0 */
+	if (!gpio_is_ready_dt(&green_gpio)) {
+		LOG_ERR("%s: device not ready.\n", green_gpio.port->name);
```

Comment:
`LOG_...` calls don't need the explicit `\n`, also this file mixes printk and LOG, I think you should stick with one or the other, if you enable logging just use LOG everywhere